libtock/usb_keyboard_hid.c

```diff
@@ -0,0 +1,232 @@
+#include <ctype.h>
+#include <stdio.h>
+
+#include "tock.h"
+#include "usb_keyboard_hid.h"
+
+struct usb_keyboard_hid_data {
+  bool fired;
+  int callback_type;
+};
+
+static struct usb_keyboard_hid_data result = { .fired = false };
+
+// Internal callback for faking synchronous reads
+static void usb_keyboard_hid_upcall(int                          callback_type,
+                                    __attribute__ ((unused)) int unused1,
+                                    __attribute__ ((unused)) int unused2,
+                                    void*                        ud) {
+  struct usb_keyboard_hid_data* data = (struct usb_keyboard_hid_data*) ud;
+  data->callback_type = callback_type;
+  data->fired         = true;
+}
+
+
+int usb_keyboard_hid_set_callback(subscribe_upcall callback, void* callback_args) {
+  subscribe_return_t sval = subscribe(DRIVER_NUM_USBKEYBOARDHID, 0, callback, callback_args);
+  return tock_subscribe_return_to_returncode(sval);
+}
+
+int usb_keyboard_hid_set_receive_buffer(uint8_t* buffer, uint32_t len) {
+  allow_rw_return_t aval = allow_readwrite(DRIVER_NUM_USBKEYBOARDHID, 1, (void*) buffer, len);
+  return tock_allow_rw_return_to_returncode(aval);
+}
+
+int usb_keyboard_hid_set_send_buffer(uint8_t* buffer, uint32_t len) {
+  allow_rw_return_t aval = allow_readwrite(DRIVER_NUM_USBKEYBOARDHID, 1, (void*) buffer, len);
+  return tock_allow_rw_return_to_returncode(aval);
+}
+
+int usb_keyboard_hid_send(void) {
+  syscall_return_t cval = command(DRIVER_NUM_USBKEYBOARDHID, 1, 0, 0);
+  return tock_command_return_novalue_to_returncode(cval);
+}
+
+int usb_keyboard_hid_send_sync(uint8_t* buffer, uint32_t len) {
```

Comment:
If `len` must be 64 bytes why not enforce that with an assert?

Reply:
Mostly because I don't _really_ want to encode this length in the API. This limitation is a by-product of how the capsule is implemented, and from an aspirational point of view, we would fix the capsule rather than have all of these specific size requirements in userland.

I wouldn't want someone to see this code and copy these rigid length expectations as that isn't how we should design syscall APIs.